lib/ex_doc/formatter/html/autolink.ex

```diff
@@ -166,26 +222,36 @@ defmodule ExDoc.Formatter.HTML.Autolink do
         {name, _, args} = form, placeholders when is_atom(name) and is_list(args) ->
           arity = length(args)
 
-          if {name, arity} in typespecs do
-            string = Macro.to_string(form)
-            n = enc_h("#{name}")
-            {string_to_link, _string_with_parens} = split_string_to_link(string)
-            string = ~s[<a href="#t:#{n}/#{arity}">#{h(string_to_link)}</a>]
-
-            put_placeholder(form, string, placeholders)
-          else
-            {form, placeholders}
+          cond do
+            {name, arity} in typespecs ->
+              n = enc_h("#{name}")
+              url = "#t:#{n}/#{arity}"
+              string = format_typespec_form(form, url)
+              put_placeholder(form, string, placeholders)
+
+            {name, arity} in @basic_types ->
+              source = get_source(Kernel, [], lib_dirs)
+              url = "#{source}#{enc_h(inspect Kernel)}.html#t:basic/0"
+              string = format_typespec_form(form, url)
+              put_placeholder(form, string, placeholders)
+
+            {name, arity} in @builtin_types ->
+              source = get_source(Kernel, [], lib_dirs)
+              url = "#{source}#{enc_h(inspect Kernel)}.html#t:builtin/0"
```

Comment:
Why `enc_h(inspect Kernel)`? Don't we know before hand what is the output of this operation?

Also wouldn't be better to consistently link to the typespecs page?